src/main/java/org/kohsuke/github/AbstractBuilder.java

```diff
@@ -0,0 +1,106 @@
+
+    /**
+     * Creates a builder.
+     *
+     * @param root
+     *            the GitHub instance to connect to.
+     * @param intermediateReturnType
+     *            the intermediate return type returned by calls to {@link #with(String, Object)}.
+     * @param builtReturnType
+     *            the final return type for built by this builder returned when {@link #done()}} is called.
+     * @param baseInstance
+     *            optional instance on which to base this builder.
+     */
+    protected AbstractBuilder(@Nonnull GitHub root,
+            @Nonnull Class<S> intermediateReturnType,
+            @Nonnull Class<R> builtReturnType,
+            @CheckForNull R baseInstance) {
+        this.requester = root.createRequest();
+        this.returnType = builtReturnType;
+        this.intermediateReturnType = intermediateReturnType;
+        this.baseInstance = baseInstance;
+        this.updateInPlace = false;
+    }
+
+    /**
+     * Finishes an update, committing changes.
+     *
+     * This method may update-in-place or not. Either way it returns the resulting instance.
+     *
+     * @return an instance with updated current data
+     * @throws IOException
+     *             if there is an I/O Exception
+     */
+    @Nonnull
+    public R done() throws IOException {
+        R result;
+        if (updateInPlace && baseInstance != null) {
+            result = requester.fetchInto(baseInstance);
+        } else {
+            result = requester.fetch(returnType);
+        }
+        return result;
+    }
+
+    /**
+     * Applies a value to a name for this builder.
+     *
+     * The internals of this method look terrifying, but they they're actually basically safe due to previous comparison
+     * of U and T determined by comparing class instances passed in during construction.
+     *
+     * If U is the same as T, this cause the builder to commit changes after the first value change and return a T from
+     * done().
+     *
+     * If U is not the same as T, the builder will batch together multiple changes and let the user call done() when
+     * they are ready.
+     *
+     * This little bit of roughness in this base class means all inheriting builders get to create Updater and Setter
+     * classes from almost identical code. Creator can be implemented with significant code reuse as well.
+     *
+     * There is probably a cleaner way to implement this, but I'm not sure what it is right now.
+     *
+     * @param name
+     *            the name of the field
+     * @param value
+     *            the value of the field
+     * @return either a continuing builder or an updated data record
+     * @throws IOException
+     *             if an I/O error occurs
+     */
+    @Nonnull
+    protected S with(@Nonnull String name, Object value) throws IOException {
+        requester.with(name, value);
+        if (returnType.equals(intermediateReturnType)) {
+            return intermediateReturnType.cast(done());
```

Comment:
This threw me off the first time I read this code to realize that this is what ensures setters still make the REST API call instead of batching

Reply:
I should make it a method or boolean field with comments to make it clearer.

Reply:
@res0nance How about now? Better?

Reply:
Much better